File: mvp/argus_roi/autopilot/engine/retrieve.py
"""Retrieve relevant chunks for a query.

  1. Embed the query (Gemini text-embedding-004).
  2. pgvector cosine ANN against ke_chunk (over-fetch k*3).
  3. Rerank by  sim × source.authority × recency_decay.
  4. Return top-k chunks with their source metadata.
"""

from __future__ import annotations

import math
from datetime import datetime, timezone

import store
from . import providers


def _emb_lit(vec: list[float]) -> str:
    return "'[" + ",".join(f"{x:.6f}" for x in vec) + "]'"


def retrieve(tenant_id: str, query: str, k: int = 8) -> list[dict]:
    """Returns: [{chunk_id, source_id, source_title, source_kind, content,
                  sim, authority, score}]."""
    if not query.strip():
        return []
    q_vec = providers.embed(tenant_id, [query])[0]
    over = max(k * 3, 12)
    rows = store.pgq(
        "SELECT c.id AS chunk_id, c.source_id, c.content, c.created_at, "
        "1 - (c.embedding <=> " + _emb_lit(q_vec) + "::vector) AS sim, "
        "s.title AS source_title, s.kind AS source_kind, s.authority "
        "FROM autopilot.ke_chunk c "
        "JOIN autopilot.ke_source s ON s.id = c.source_id "
        f"WHERE c.tenant_id = {store.dq(tenant_id)} "
        f"ORDER BY c.embedding <=> {_emb_lit(q_vec)}::vector "
        f"LIMIT {int(over)};")

    now = datetime.now(timezone.utc)
    scored = []
    for r in rows:
        sim = float(r["sim"])
        auth = float(r["authority"])
        # mild recency boost over the last year
        try:
            age_days = (now - datetime.fromisoformat(r["created_at"].replace("Z", "+00:00"))).days
        except Exception:
            age_days = 0
        recency = math.exp(-age_days / 365.0)   # 1.0 today, ~0.37 at 1y
        score = sim * (0.5 + 0.5 * auth) * (0.7 + 0.3 * recency)
        scored.append({
            "chunk_id": int(r["chunk_id"]),
            "source_id": int(r["source_id"]),
            "source_title": r["source_title"],
            "source_kind": r["source_kind"],
            "content": r["content"],
            "sim": round(sim, 4),
            "authority": auth,
            "score": round(score, 4),
        })
    scored.sort(key=lambda x: -x["score"])
    return scored[:k]

File: apps/api/src/llm/retrieve.ts
/**
 * Retrieve relevant chunks for a query.
 *
 * 1. Embed the query (Gemini).
 * 2. pgvector cosine ANN against `chunks` (over-fetch k*3).
 * 3. Rerank by  sim × (0.5 + 0.5·authority) × (0.7 + 0.3·recency_decay).
 *    Authority is 0..100 (smallint in `sources`), normalised to 0..1.
 *    Recency uses exp(-age_days / 365), so today=1.0, ~0.37 at one year.
 * 4. Return top-k with source metadata.
 */
import { db } from '../db.js';
import { embed, type ProviderRow } from './gemini.js';
import { toPgvectorLiteral } from './pgvector.js';

export interface RetrievedChunk {
  chunk_id: string;
  source_id: string;
  source_title: string;
  source_kind: string;
  content: string;
  sim: number;
  authority: number;
  score: number;
}

interface RawRow {
  chunk_id: string;
  source_id: string;
  content: string;
  created_at: string;
  sim: string; // pg returns the numeric expression as a string by default
  source_title: string;
  source_kind: string;
  authority: number;
}

export async function retrieve(
  envId: string,
  query: string,
  provider: ProviderRow,
  k = 8,
): Promise<RetrievedChunk[]> {
  if (!query.trim()) return [];
  const [qVec] = await embed(provider, [query], 768);
  if (!qVec) return [];
  const literal = toPgvectorLiteral(qVec);
  const over = Math.max(k * 3, 12);

  const { rows } = await db().query<RawRow>(
    `SELECT c.id            AS chunk_id,
            c.source_id     AS source_id,
            c.text          AS content,
            c.created_at,
            1 - (c.embedding <=> $1::vector) AS sim,
            s.title         AS source_title,
            s.kind          AS source_kind,
            s.authority     AS authority
       FROM chunks c
       JOIN sources s ON s.id = c.source_id
      WHERE c.env_id = $2
      ORDER BY c.embedding <=> $1::vector
      LIMIT $3`,
    [literal, envId, over],
  );

  const now = Date.now();
  const scored: RetrievedChunk[] = rows.map((r) => {
    const sim = Number(r.sim);
    const authNorm = Math.max(0, Math.min(1, r.authority / 100));
    const ageDays = Math.max(0, (now - new Date(r.created_at).getTime()) / 86_400_000);
    const recency = Math.exp(-ageDays / 365);
    const score = sim * (0.5 + 0.5 * authNorm) * (0.7 + 0.3 * recency);
    return {
      chunk_id: r.chunk_id,
      source_id: r.source_id,
      source_title: r.source_title,
      source_kind: r.source_kind,
      content: r.content,
      sim: Number(sim.toFixed(4)),
      authority: r.authority,
      score: Number(score.toFixed(4)),
    };
  });
  scored.sort((a, b) => b.score - a.score);
  return scored.slice(0, k);
}
